### PR TITLE
Infrastructure for confidential identities

### PR DESCRIPTION
**Pull request by rnicoll**
* De-anonymise parties in AbstractStateReplacementFlow flows
* Convert transaction key negotiation to a subflow instead of utility functions
* Add serialization support for `CertPath` and `X509Certificate`
* Restructure cash flows so that a counterparty flow can be added later

### CODE REVIEW

**Comment by rnicoll** [comment]
This is infrastructure changes for #714, in an attempt to make that PR a little less invasive

File: core/src/main/kotlin/net/corda/core/flows/TxKeyFlow.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+        companion object {
+            object AWAITING_KEY : ProgressTracker.Step("Awaiting key")
+
+            fun tracker() = ProgressTracker(AWAITING_KEY)
+        }
+
+        @Suspendable
+        override fun call(): AnonymousIdentity {
+            progressTracker.currentStep = AWAITING_KEY
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/flows/TxKeyFlow.kt]: I think this has to be an exchange, as there is no way to push our key to them in advance of their verify. Using Issuance is probably not a good first choice as Move is probably more general.

**Comment by rnicoll** [comment]
Restructured the transaction key flow to send identities in both directions and return a map. Also means that extending it to multi-way identity exchange shouldn't be too hard.

**Comment by rnicoll** [comment]
Ripped out the flow name changes pending upcoming work from Shams, which should make this rather a lot easier to read

**Comment by rnicoll** [comment]
Rebased off Patrick's changes and squashed down, ready for re-review

File: core/src/main/kotlin/net/corda/core/node/services/Services.kt [hunk 2]
```diff
@@ -385,6 +387,24 @@ interface KeyManagementService {
     @Suspendable
     fun freshKey(): PublicKey
 
+    /**
+     * Generates a new random [KeyPair], adds it to the internal key storage, then generates a corresponding
+     * [X509Certificate] and adds it to the identity service. Returns the X.509 certificate.
+     */
+    @Suspendable
+    fun freshKeyAndCert(identity: Party, revocationEnabled: Boolean): Pair<X509Certificate, CertPath> {
+        val ourPublicKey = freshKey()
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/node/services/Services.kt]: Please make this a bland interface line, with implementation in node. This ties us to a specific implementation.

**Comment by rnicoll** [comment]
Sorry in trying to get TC to re-run I've lost your review; can you re-accept please

**Pull request merged**